File: src/spacetraders/adapters/secondary/routing/graph_builder.py
"""Graph building adapter - constructs navigation graphs from API data"""
import logging
import math
from typing import Dict, List

from ....ports.outbound.api_client import ISpaceTradersAPI
from ....ports.outbound.graph_provider import IGraphBuilder

logger = logging.getLogger(__name__)


def euclidean_distance(x1: float, y1: float, x2: float, y2: float) -> float:
    """Calculate Euclidean distance between two coordinates"""
    return math.hypot(x2 - x1, y2 - y1)


class GraphBuilder(IGraphBuilder):
    """Builds system navigation graphs by fetching waypoints from the API"""

    def __init__(self, api_client: ISpaceTradersAPI):
        self.api = api_client

    def build_system_graph(self, system_symbol: str) -> Dict:
        """
        Fetch all waypoints for system and build navigation graph

        Returns:
            Graph dict: {waypoints: {symbol: {...}}, edges: [{from, to, distance, type}]}
        """
        logger.info(f"Building graph for system {system_symbol}...")

        # Fetch all waypoints with pagination
        all_waypoints: List[Dict] = []
        page = 1
        limit = 20

        while True:
            try:
                result = self.api.list_waypoints(system_symbol, limit=limit, page=page)
            except Exception as e:
                logger.error(f"Failed to fetch waypoints page {page}: {e}")
                raise RuntimeError(f"API error while building graph for {system_symbol}") from e

            if not result or "data" not in result:
                break

            waypoints_page = result["data"]
            all_waypoints.extend(waypoints_page)

            logger.info(f"  Fetched page {page}: {len(waypoints_page)} waypoints")

            # Check if we have more pages
            meta = result.get("meta", {})
            total = meta.get("total", 0)
            total_pages = (total // limit) + (1 if total % limit > 0 else 0)

            if page >= total_pages or len(waypoints_page) < limit:
                break

            page += 1

            # Safety limit
            if page > 50:
                logger.warning("Reached safety limit of 50 pages")
                break

        if not all_waypoints:
            logger.error(f"No waypoints found for system {system_symbol}")
            raise RuntimeError(f"No waypoints found for system {system_symbol}")

        # Build graph structure
        graph = {
            "system": system_symbol,
            "waypoints": {},
            "edges": [],
        }

        # Process waypoints
        for waypoint in all_waypoints:
            traits = [t["symbol"] for t in waypoint.get("traits", [])]
            has_fuel = "MARKETPLACE" in traits or "FUEL_STATION" in traits

            graph["waypoints"][waypoint["symbol"]] = {
                "type": waypoint.get("type"),
                "x": waypoint.get("x"),
                "y": waypoint.get("y"),
                "traits": traits,
                "has_fuel": has_fuel,
                "orbitals": [o["symbol"] for o in waypoint.get("orbitals", [])],
            }

        # Build edges (bidirectional graph)
        waypoint_list = list(graph["waypoints"].keys())

        for i, wp1 in enumerate(waypoint_list):
            wp1_data = graph["waypoints"][wp1]

            # Only create edges with waypoints that come after this one (avoid duplicates)
            for wp2 in waypoint_list[i + 1:]:
                wp2_data = graph["waypoints"][wp2]

                # Check if this is an orbital relationship (zero distance)
                is_orbital = (
                    wp2 in wp1_data.get("orbitals", []) or
                    wp1 in wp2_data.get("orbitals", [])
                )

                if is_orbital:
                    distance = 0.0
                    edge_type = "orbital"
                else:
                    distance = euclidean_distance(
                        wp1_data["x"], wp1_data["y"],
                        wp2_data["x"], wp2_data["y"],
                    )
                    edge_type = "normal"

                distance = round(distance, 2)

                # Add bidirectional edges
                graph["edges"].append({
                    "from": wp1,
                    "to": wp2,
                    "distance": distance,
                    "type": edge_type,
                })
                graph["edges"].append({
                    "from": wp2,
                    "to": wp1,
                    "distance": distance,
                    "type": edge_type,
                })

        logger.info(f"Graph built for {system_symbol}")
        logger.info(f"  Waypoints: {len(graph['waypoints'])}")
        logger.info(f"  Edges: {len(graph['edges'])}")
        logger.info(
            f"  Fuel stations: {sum(1 for wp in graph['waypoints'].values() if wp['has_fuel'])}"
        )

        return graph

File: gobot/internal/adapters/api/graph_builder.go
package api

import (
	"context"
	"fmt"
	"log"
	"math"
	"slices"

	"github.com/andrescamacho/spacetraders-go/internal/domain/player"
	domainPorts "github.com/andrescamacho/spacetraders-go/internal/domain/ports"
	"github.com/andrescamacho/spacetraders-go/internal/domain/shared"
	"github.com/andrescamacho/spacetraders-go/internal/domain/system"
)

// maxWaypointPages caps waypoint pagination so a server that never reports the
// collection exhausted cannot spin the fetch loop forever.
const maxWaypointPages = 50

// GraphBuilder builds system navigation graphs from API data
type GraphBuilder struct {
	apiClient    domainPorts.APIClient
	playerRepo   player.PlayerRepository
	waypointRepo system.WaypointRepository
}

// NewGraphBuilder creates a new graph builder
func NewGraphBuilder(
	apiClient domainPorts.APIClient,
	playerRepo player.PlayerRepository,
	waypointRepo system.WaypointRepository,
) system.IGraphBuilder {
	return &GraphBuilder{
		apiClient:    apiClient,
		playerRepo:   playerRepo,
		waypointRepo: waypointRepo,
	}
}

// BuildSystemGraph fetches all waypoints from API and builds navigation graph with dual-cache strategy
//
// Populates both:
// 1. Return value: NavigationGraph for navigation (infinite TTL)
// 2. Waypoints table: Full trait data for queries (2hr TTL)
func (b *GraphBuilder) BuildSystemGraph(ctx context.Context, systemSymbol string, playerID int) (*system.NavigationGraph, error) {
	log.Printf("Building graph for system %s...", systemSymbol)

	playerIDValue, err := shared.NewPlayerID(playerID)
	if err != nil {
		return nil, fmt.Errorf("invalid player ID: %w", err)
	}
	player, err := b.playerRepo.FindByID(ctx, playerIDValue)
	if err != nil {
		return nil, fmt.Errorf("failed to get player: %w", err)
	}

	allWaypoints, err := b.fetchAllWaypoints(ctx, systemSymbol, player.Token)
	if err != nil {
		return nil, err
	}

	if len(allWaypoints) == 0 {
		return nil, fmt.Errorf("no waypoints found for system %s", systemSymbol)
	}

	graph := system.NewNavigationGraph(systemSymbol)

	waypointObjects := []*shared.Waypoint{}

	for _, wp := range allWaypoints {
		waypointObj, err := waypointFromAPIData(wp, systemSymbol)
		if err != nil {
			log.Printf("Warning: failed to create waypoint %s: %v", wp.Symbol, err)
			continue
		}

		graph.AddWaypoint(waypointObj)

		waypointObjects = append(waypointObjects, waypointObj)
	}

	connectEveryWaypointPair(graph)
	b.cacheWaypoints(ctx, waypointObjects)

	fuelStations := len(graph.GetFuelStations())

	log.Printf("Graph built for %s", systemSymbol)
	log.Printf("  Waypoints: %d", graph.WaypointCount())
	log.Printf("  Edges: %d", graph.EdgeCount())
	log.Printf("  Synced %d waypoints to waypoints table", len(waypointObjects))
	log.Printf("  Fuel stations: %d", fuelStations)

	return graph, nil
}

func (b *GraphBuilder) fetchAllWaypoints(ctx context.Context, systemSymbol, token string) ([]system.WaypointAPIData, error) {
	allWaypoints := []system.WaypointAPIData{}
	limit := apiPageLimitMax

	for page := 1; page <= maxWaypointPages; page++ {
		result, err := b.apiClient.ListWaypoints(ctx, systemSymbol, token, page, limit)
		if err != nil {
			return nil, fmt.Errorf("failed to fetch waypoints page %d: %w", page, err)
		}

		if len(result.Data) == 0 {
			break
		}

		allWaypoints = append(allWaypoints, result.Data...)

		log.Printf("  Fetched page %d: %d waypoints", page, len(result.Data))

		totalPages := (result.Meta.Total / limit)
		if result.Meta.Total%limit > 0 {
			totalPages++
		}

		if page >= totalPages || len(result.Data) < limit {
			break
		}
	}

	return allWaypoints, nil
}

func waypointFromAPIData(wp system.WaypointAPIData, systemSymbol string) (*shared.Waypoint, error) {
	waypointObj, err := shared.NewWaypoint(wp.Symbol, wp.X, wp.Y)
	if err != nil {
		return nil, err
	}

	traits := traitSymbols(wp.Traits)

	waypointObj.SystemSymbol = systemSymbol
	waypointObj.Type = wp.Type
	waypointObj.Traits = traits
	waypointObj.HasFuel = shared.WaypointGrantsFuel(wp.Type, traits)
	waypointObj.Orbitals = orbitalSymbols(wp.Orbitals)

	return waypointObj, nil
}

func orbitalSymbols(orbitals []map[string]string) []string {
	symbols := []string{}
	for _, orbital := range orbitals {
		if symbol, ok := orbital["symbol"]; ok {
			symbols = append(symbols, symbol)
		}
	}
	return symbols
}

func traitSymbols(traits []map[string]interface{}) []string {
	symbols := []string{}
	for _, trait := range traits {
		if symbol, ok := trait["symbol"]; ok {
			if symbolStr, ok := symbol.(string); ok {
				symbols = append(symbols, symbolStr)
			}
		}
	}
	return symbols
}

func connectEveryWaypointPair(graph *system.NavigationGraph) {
	waypointList := make([]string, 0, len(graph.Waypoints))
	for symbol := range graph.Waypoints {
		waypointList = append(waypointList, symbol)
	}

	for i, wp1Symbol := range waypointList {
		wp1 := graph.Waypoints[wp1Symbol]

		// Only create edges with waypoints that come after this one (avoid duplicates)
		for _, wp2Symbol := range waypointList[i+1:] {
			distance, edgeType := edgeBetween(wp1, graph.Waypoints[wp2Symbol])
			graph.AddEdge(wp1Symbol, wp2Symbol, distance, edgeType)
		}
	}
}

func edgeBetween(wp1, wp2 *shared.Waypoint) (float64, system.EdgeType) {
	if inSameOrbit(wp1, wp2) {
		return 0.0, system.EdgeTypeOrbital
	}
	return math.Round(wp1.DistanceTo(wp2)*100) / 100, system.EdgeTypeNormal
}

func inSameOrbit(wp1, wp2 *shared.Waypoint) bool {
	return slices.Contains(wp1.Orbitals, wp2.Symbol) || slices.Contains(wp2.Orbitals, wp1.Symbol)
}

func (b *GraphBuilder) cacheWaypoints(ctx context.Context, waypointObjects []*shared.Waypoint) {
	for _, waypointObj := range waypointObjects {
		if err := b.waypointRepo.Add(ctx, waypointObj); err != nil {
			log.Printf("Warning: failed to save waypoint %s: %v", waypointObj.Symbol, err)
			// Continue - caching failure shouldn't break the operation
		}
	}
}
